R/filters.R

```diff
@@ -1,70 +1,115 @@
 #' data(immdata)
 #'
 #' # Select samples with status "MS"
-#' repFilter(immdata, "by.meta", list(Status = "MS"))
+#' repFilter(immdata, "by.meta", list(Status = include("MS")))
+#'
+#' # Select samples without status "MS"
+#' repFilter(immdata, "by.meta", list(Status = exclude("MS")))
+#'
+#' # Select samples from lanes "A" and "B" with age > 15
+#' repFilter(immdata, "by.meta", list(Lane = include("A", "B"), Age = morethan(15)))
+#'
+#' # Select samples that are not from lanes "A" and "B"
+#' repFilter(immdata, "by.meta", list(Lane = exclude("A", "B")))
 #'
-#' # Select samples from lanes "A" and "B" with age >= 15
-#' repFilter(immdata, "by.meta", list(Lane = c("A", "B"), Age = ">=15"))
+#' # Select samples with a number of clonotypes from 1000 to 5000
+#' repFilter(immdata, "by.repertoire", list(n_clonotypes = interval(1000, 5000)))
+#'
+#' # Select clonotypes in all samples with alpha chains
+#' repFilter(immdata, "by.clonotype",
+#'   list(V.name = include("AV"), J.name = include("AJ")),
+#'   match = "substring"
+#' )
 #' @export repFilter
-repFilter <- function(.data, .method, .query, ...) {
+repFilter <- function(.data, .method = "by.clonotype",
+                      .query = list(CDR3.aa = exclude("partial", "out_of_frame")),
+                      .match = "exact") {
   if (!is.list(.data)) {
     stop(paste0(
       "Input data is not a list; ",
-      "please pass Immunarch object with data and metadata as input."
+      "please pass Immunarch dataset object as input."
     ))
   }
-  if (!("meta" %in% names(.data))) {
+
+  if (!(.match %in% c("exact", "startswith", "substring"))) {
     stop(paste0(
-      "Input data doesn't contain meta; ",
-      "please pass Immunarch object with data and metadata as input."
+      "Unknown matching method \"", .match,
+      "\"! Supported matching methods are: ",
+      "\"exact\", \"startswith\", \"substring\"."
     ))
   }
 
   switch(tolower(.method),
-    by.meta = filter_by_meta(.data, .query),
-    stop("You entered the wrong method! Supported methods are: \"by.meta\".")
+    by.meta = filter_by_meta(.data, .query, .match),
+    by.repertoire = filter_by_repertoire(.data, .query),
+    by.rep = filter_by_repertoire(.data, .query),
+    by.clonotype = filter_by_clonotype(.data, .query, .match),
+    by.cl = filter_by_clonotype(.data, .query, .match),
+    stop(paste0(
+      "You entered wrong method \"", .method, "\"! Supported methods are: ",
+      "\"by.meta\", \"by.repertoire\", \"by.clonotype\"."
+    ))
   )
 }
 
-filter_by_meta <- function(.data, .query) {
-  names <- names(.query)
+filter_by_meta <- function(.data, .query, .match) {
+  if (!("meta" %in% names(.data))) {
+    stop(paste0(
+      "Input data doesn't contain meta; ",
+      "please pass Immunarch dataset object with data and metadata as input."
+    ))
+  }
 
   filtered_meta <- .data$meta
-  for (i in seq_along(names)) {
-    name <- names[i]
+  for (i in seq_along(names(.query))) {
+    name <- names(.query)[i]
```

Comment:
if (length(names(.query))==0){stop ('Unnamed list could not be passed as query, please provide named list')}
 for (i in 1:length(.query)) {
    name <- names(.query)[i]
    if (name==''') {stop(All values in list should be named)} 
```suggestion
 if (length(names(.query))==0){stop ('Unnamed list could not be passed as query, please provide named list')}
 for (i in 1:length(.query)) {
    name <- names(.query)[i]
    if (name==''') {stop(All values in list should be named)} 
```

Reply:
Better place for query validation is in the beginning of repFilter(); I will add it there